README.md
# My Wishlist
My Wishlist is a wishlist AddOn dedicated to raiders wanting to prepare their item priorities. Toggle UI through minimap icon on `/mwl`. `[Not currently present:]` AddOn allows for configuration of popular loot auction addons integration, to automatically popup when loot distribution starts.

# Features
* Adding items through slash commands (`/prio`) and UI
* Tooltip display
* Per-slot wishlists
* Togglable lock/unlock
* Profiles
* Adjustable UI

# Planned features
## Loot distribution AddOns Integrations
* Classic Loot Manager
* Gargul
* RCLootCouncil

## TMB Wishlist Import

## Other players wishlist inspection
## Configuration
## Localization


CHANGELOG.md
# My Wishlist v0.1.0
## Initial release of Core functionality

MyWishlist.lua
local name, MWL = ...

local Core = LibStub("AceAddon-3.0"):NewAddon(name, "AceEvent-3.0")

MWL.InternalSlots = {
    ["Head"] = "Head", ["Neck"] = "Neck", ["Shoulders"] = "Shoulders", ["Back"] = "Back", ["Chest"] = "Chest", ["Wrist"] = "Wrist", ["Hands"] = "Hands", ["Waist"] = "Waist", ["Legs"] = "Legs", ["Feet"] = "Feet", ["Finger"] = "Finger",
    ["Trinket"] = "Trinket", ["Weapon"] = "Weapon", ["OffHand"] = "Held In Off-Hand", ["Ranged"] = "Ranged", ["Ammo"] = "Ammo", ["Shirt"] = "Shirt", ["Tabard"] = "Tabard", ["Miscellaneous"] = "Miscellaneous"
}

MWL.InternalSlotsOrdered = {
    "Head", "Neck", "Shoulders", "Back", "Chest", "Wrist", "Hands", "Waist", "Legs", "Feet", "Finger",
    "Trinket", "Weapon", "OffHand", "Ranged", "Ammo", "Shirt", "Tabard", "Miscellaneous"
}



local defaults = {
    global = {
        version = {
            major = 0,
            minor = 0,
            patch = 0,
            changeset = "000000"
        },
        versionString = "v0.0.0-000000",
        lock = false
    },
    profile = {
        wishlists = {},
        minimap = {},
        gui = {
            height = 380
        }
    }
}
do
    local function UpdateDefaults()
        for slot,_ in pairs(MWL.InternalSlots) do
            defaults.profile.wishlists[slot] = {}
        end
    end

    UpdateDefaults()
end

function Core:OnInitialize()
    self.db = LibStub("AceDB-3.0"):New("MWL_DB", defaults, true)

    local newVersion = MWL.Utils.GetVersion("@project-version@")

    if MWL.Utils.CompareVersions(newVersion, self.db.global.version) then
        -- Handle version change code
    end

    self.db.global.version = newVersion
    self.db.global.versionString = MWL.Utils.GetVersionString(newVersion)

    MWL.ChatCommands:Register()
end

function Core:OnEnable()
    MWL.Manager:Initialize()
    -- MWL.Settings:Initialize()
    MWL.GUI:Initialize()
    MWL.Minimap:Initialize()
end

function Core:GetVersion()
    return self.db.global.version
end

function Core:GetVersionString()
    return self.db.global.versionString
end

MWL.Core = Core

source/Utils.lua
local _, MWL = ...

local smatch, sformat, sfind = string.match, string.format, string.find
local tonumber = tonumber
local strsub = strsub

local Utils = {}

function Utils.GetVersion(versionString)
    local major, minor, patch, changeset = smatch(versionString, "^v(%d+).(%d+).(%d+)-?(.*)")
    return {
        major = tonumber(major) or 0,
        minor = tonumber(minor) or 0,
        patch = tonumber(patch) or 0,
        changeset = changeset or ""
    }
end

function Utils.GetVersionString(version)
    local changeset = ""
    if version.changeset and version.changeset ~= "" then
        changeset = "-" .. changeset
    end
    return sformat(
        "v%s.%s.%s%s",
        version.major or 0,
        version.minor or 0,
        version.patch or 0,
        changeset)
end

local function VersionToNumber(v)
    return v.major*10000 + v.minor*100 + v.patch
end

function Utils.CompareVersions(v1, v2)
    return VersionToNumber(v1) > VersionToNumber(v2)
end

function Utils.GetItemIdFromLink(itemLink)
    -- local _, _, Color, Ltype, Id, Enchant, Gem1, Gem2, Gem3, Gem4, Suffix, Unique, LinkLvl, Name = sfind(itemLink, "|?c?f?f?(%x*)|?H?([^:]*):?(%d+):?(%d*):?(%d*):?(%d*):?(%d*):?(%d*):?(%-?%d*):?(%-?%d*):?(%d*):?(%d*):?(%-?%d*)|?h?%[?([^%[%]]*)%]?|?h?|?r?")
    itemLink = itemLink or ""
    local _, _, _, _, itemId = sfind(itemLink, "|?c?f?f?(%x*)|?H?([^:]*):?(%d+).*")
    return tonumber(itemId)
end


function Utils.strsub20(s)
    return strsub(s, 1, 20)
end

function Utils.ColorCodeText(text, color)
    return sformat("|cff%s%s|r", color, text)
end

MWL.Utils = Utils

source/Manager.lua
local _,  MWL = ...

local pairs, ipairs, type = pairs, ipairs, type
local Item, GetItemInfoInstant = Item, GetItemInfoInstant
local tinsert, tremove = table.insert, table.remove

local Manager = {}

local INVTYPE_to_MWL_slot_map = {
    ["INVTYPE_HEAD"]            = MWL.InternalSlots.Head,
    ["INVTYPE_NECK"]            = MWL.InternalSlots.Neck,
    ["INVTYPE_SHOULDER"]        = MWL.InternalSlots.Shoulders,
    ["INVTYPE_BODY"]            = MWL.InternalSlots.Chest,
    ["INVTYPE_CHEST"]           = MWL.InternalSlots.Chest,
    ["INVTYPE_WAIST"]           = MWL.InternalSlots.Waist,
    ["INVTYPE_LEGS"]            = MWL.InternalSlots.Legs,
    ["INVTYPE_FEET"]            = MWL.InternalSlots.Feet,
    ["INVTYPE_WRIST"]           = MWL.InternalSlots.Wrist,
    ["INVTYPE_HAND"]            = MWL.InternalSlots.Hand,
    ["INVTYPE_FINGER"]          = MWL.InternalSlots.Finger,
    ["INVTYPE_TRINKET"]         = MWL.InternalSlots.Trinket,
    ["INVTYPE_WEAPON"]          = MWL.InternalSlots.Weapon,
    ["INVTYPE_SHIELD"]          = MWL.InternalSlots.OffHand,
    ["INVTYPE_RANGED"]          = MWL.InternalSlots.Ranged,
    ["INVTYPE_CLOAK"]           = MWL.InternalSlots.Back,
    ["INVTYPE_2HWEAPON"]        = MWL.InternalSlots.Weapon,
    ["INVTYPE_TABARD"]          = MWL.InternalSlots.Tabard,
    ["INVTYPE_ROBE"]            = MWL.InternalSlots.Chest,
    ["INVTYPE_WEAPONMAINHAND"]  = MWL.InternalSlots.Weapon,
    ["INVTYPE_WEAPONOFFHAND"]   = MWL.InternalSlots.Weapon,
    ["INVTYPE_HOLDABLE"]        = MWL.InternalSlots.OffHand,
    ["INVTYPE_THROWN"]          = MWL.InternalSlots.Ranged,
    ["INVTYPE_RANGEDRIGHT"]     = MWL.InternalSlots.Ranged,
    ["INVTYPE_RELIC"]           = MWL.InternalSlots.Ranged,
}

-- local INVSLOT_REDUCTION = {
--     [INVSLOT_FINGER2] = INVSLOT_FINGER1,
--     [INVSLOT_TRINKET2] = INVSLOT_TRINKET1
-- }

local function UpdateWishlistedItemMetadata(self, id, add)
    if not self.wishlistedItemMap[id] then
        self.wishlistedItemMap[id] = 0

        -- Slot
        local _, _, _, itemEquipLoc, _, _, _ = GetItemInfoInstant(id)
        local slot = INVTYPE_to_MWL_slot_map[itemEquipLoc] or MWL.InternalSlots.Miscellaneous
        self.wishlistedItemSlotMap[id] = slot
    end
    if add then
        self.wishlistedItemMap[id] = self.wishlistedItemMap[id] + 1
    else
        self.wishlistedItemMap[id] = self.wishlistedItemMap[id] - 1
        if self.wishlistedItemMap[id] < 0 then
            error("Wishlisted item count went below 0", 2)
        end
    end
end

local function BuildWishlistsFromDatabase(self)
    self.wishlists = {}
    self.wishlistedItemMap = {}
    self.wishlistedItemSlotMap = {}
    for slot, _ in pairs(MWL.InternalSlots) do
        self.wishlists[slot] = {}
        self.wishlistedItemMap[slot] = {}
        for _, entry in ipairs(MWL.Core.db.profile.wishlists[slot]) do
            local item = Item:CreateFromItemID(entry.id)
            if not item:IsItemEmpty() then
                UpdateWishlistedItemMetadata(self, entry.id, true)
                self.wishlists[slot][#self.wishlists[slot]+1] = MWL.NewWishlistEntry(item, entry.note)
            end
        end
    end
end

local function Store()
    for slot, _ in pairs(MWL.InternalSlots) do
        MWL.Core.db.profile.wishlists[slot] = {}
        for _, entry in ipairs(Manager.wishlists[slot]) do
            if not entry:IsItemEmpty() then
                MWL.Core.db.profile.wishlists[slot][#MWL.Core.db.profile.wishlists[slot]+1] = {
                    id = tonumber(entry:GetItemID()) or 0,
                    note = tostring(entry:GetNote()) or ""
                }
            end
        end
    end
end

local function RegisterDBCallbacks(self)
    MWL.Core.db.RegisterCallback(self, "OnDatabaseShutdown", Store)
end

function Manager:Initialize()
    BuildWishlistsFromDatabase(self)
    RegisterDBCallbacks(self)
end

local function AddItemInternal(self, itemId, note, position)
    if self:IsLocked() then return end
    local item = Item:CreateFromItemID(itemId)
    if item:IsItemEmpty() then return end

    local _, _, _, itemEquipLoc, _, _, _ = GetItemInfoInstant(itemId)

    local slot = INVTYPE_to_MWL_slot_map[itemEquipLoc] or MWL.InternalSlots.Miscellaneous
    local entry = MWL.NewWishlistEntry(item, note)
    tinsert(self.wishlists[slot], position, entry)
    UpdateWishlistedItemMetadata(self, itemId, true)
end

function Manager:AddItemById(itemId, note, position)
    if self:IsLocked() then return end
    itemId = tonumber(itemId) or 0
    if not GetItemInfoInstant(itemId) then return end
    if type(note) ~= "string" then
        note = ""
    end
    position = tonumber(position) or 1
    AddItemInternal(self, itemId, note, position)
end

function Manager:AddItemByLink(itemLink, note, position)
    if self:IsLocked() then return end
    local itemId = MWL.Utils.GetItemIdFromLink(itemLink)
    if not itemId then return end
    if not GetItemInfoInstant(itemId) then return end
    if type(note) ~= "string" then
        note = ""
    end
    position = tonumber(position) or 1
    AddItemInternal(self, itemId, note, position)
end

function Manager:GetWishlist(slot)
    return self.wishlists[slot] or {}
end

function Manager:MoveWishlistItemUp(slot, seqId)
    if self:IsLocked() then return end
    local entry = tremove(self.wishlists[slot], seqId)
    local newSeq = seqId - 1
    tinsert(self.wishlists[slot], newSeq, entry)
end

function Manager:MoveWishlistItemDown(slot, seqId)
    if self:IsLocked() then return end
    local entry = tremove(self.wishlists[slot], seqId)
    tinsert(self.wishlists[slot], seqId + 1, entry)
end

function Manager:RemoveWishlistItem(slot, seqId)
    if self:IsLocked() then return end
    local entry = tremove(self.wishlists[slot], seqId)
    UpdateWishlistedItemMetadata(self, entry:GetItemID(), false)
end

function Manager:IsItemWishlisted(itemId)
    return self.wishlistedItemMap[itemId] and self.wishlistedItemMap[itemId] > 0
end

function Manager:GetWishlistedItemSlot(itemId)
    return self.wishlistedItemSlotMap[itemId]
end

function Manager:GetWishlistedItems(itemId)
    local items = {}
    if not self:IsItemWishlisted(itemId) then return items end
    if not self:GetWishlistedItemSlot(itemId) then return items end

    for _,entry in ipairs(self.wishlists[self:GetWishlistedItemSlot(itemId)]) do
        if entry:GetItemID() == itemId then
            items[#items+1] = entry
        end
    end

    return items
end

function Manager:IsLocked()
    return MWL.Core.db.global.lock
end

function Manager:Lock()
    MWL.Core.db.global.lock = true
end

function Manager:Unlock()
    MWL.Core.db.global.lock = false
end

function Manager:ToggleLock()
    if self:IsLocked() then
        self:Unlock()
    else
        self:Lock()
    end
end

MWL.Manager = Manager

source/MinimapIcon.lua
local _, MWL = ...

local addonName = "My Wishlist" -- same as the UI name for config

local ldb = LibStub:GetLibrary("LibDataBroker-1.1", true)
if not ldb then return end

local icon = LibStub("LibDBIcon-1.0", true)
if not icon then return end

local function getIcon()
    local iconPath = "interface/addons/mywishlist/media/unlocked_3"
    if MWL.Manager:IsLocked() then
        iconPath = "interface/addons/mywishlist/media/locked_3"
    end
    return iconPath
end

local function CreateMinimapDBI(self, dropdown)
    MWL.MinimapDBI = ldb:NewDataObject(addonName, {
        type = "data source",
        text = "0",
        icon = getIcon()
    })

    MWL.MinimapDBI.OnClick = function(s, button)
        if button == "RightButton" then
            MWL.Manager:ToggleLock()
            MWL.GUI:Refresh()
            icon.objects[addonName]:GetScript("OnLeave")(icon.objects[addonName])
            icon.objects[addonName]:GetScript("OnEnter")(icon.objects[addonName])
        else
            MWL.GUI:Toggle()
        end
    end

    MWL.MinimapDBI.OnTooltipShow = function(tooltip)
        tooltip:AddDoubleLine(addonName, MWL.Core:GetVersionString())

        if MWL.Manager:IsLocked() then
            tooltip:AddDoubleLine(" ", MWL.Utils.ColorCodeText("Locked", "ee4444"))
        else
            tooltip:AddDoubleLine(" ", MWL.Utils.ColorCodeText("Unlocked", "44ee44"))
        end

        tooltip:AddDoubleLine("Left Click",            MWL.Utils.ColorCodeText("Toggle window", "bbbbbb"))
        tooltip:AddDoubleLine("Right Click",           MWL.Utils.ColorCodeText("Toggle lock", "bbbbbb"))
        tooltip:AddDoubleLine("Shift + Left Click",    MWL.Utils.ColorCodeText("Toggle options", "bbbbbb"))
    end

    icon:Register(addonName, MWL.MinimapDBI, MWL.Core.db.profile.minimap)
end

-- local dropdown
local Minimap = {}

function Minimap:Initialize()
    -- Create Minimap Icon
    CreateMinimapDBI(self, nil)

    -- Hook Minimap Icon
    hooksecurefunc(MWL.Manager, "Lock", function()
        MWL.MinimapDBI.icon = getIcon()
    end)

    hooksecurefunc(MWL.Manager, "Unlock", function()
        MWL.MinimapDBI.icon = getIcon()
    end)

    hooksecurefunc(MWL.Manager, "ToggleLock", function()
        MWL.MinimapDBI.icon = getIcon()
    end)

    -- if CLM2_MinimapIcon.disable then icon:Hide(addonName) end

end


function Minimap:Enable()
    -- CLM2_MinimapIcon.disable = false
    icon:Show(addonName)
end

function Minimap:Disable()
    -- CLM2_MinimapIcon.disable = true
    icon:Hide(addonName)
end

function Minimap:IsEnabled()
    -- return not CLM2_MinimapIcon.disable
end

MWL.Minimap = Minimap

source/ChatCommands.lua
local _,  MWL = ...

local ChatCommands = {}

local AceConsole = LibStub("AceConsole-3.0")


local function HandlePrio(itemLink)
    local itemId = MWL.Utils.GetItemIdFromLink(itemLink)
    if not itemId then
        -- TODO HANDLE ERROR
        return
    end

    MWL.Manager:AddItemById(itemId, "Slash CMD")
end

local function HandleDeprio(itemLink)
    local itemId = MWL.Utils.GetItemIdFromLink(itemLink)
    if not itemId then
        -- TODO HANDLE ERROR
        return
    end
end

local function HandleMwl()
    MWL.GUI:Toggle()
end

function ChatCommands:Register()
    AceConsole:RegisterChatCommand("prio", HandlePrio)
    AceConsole:RegisterChatCommand("deprio", HandleDeprio)
    AceConsole:RegisterChatCommand("mwl", HandleMwl)
    AceConsole:RegisterChatCommand("mywishlist", HandleMwl)
    AceConsole:RegisterChatCommand("wishlist", HandleMwl)
end

MWL.ChatCommands = ChatCommands

source/WishlistEntry.lua
local _, MWL = ...

local tostring = tostring

local WishlistEntry = {} -- WishlistEntry
WishlistEntry.__index = WishlistEntry

local strsub20 = MWL.Utils.strsub20

local function LimitNote(note)
    return strsub20(tostring(note or ""))
end

function WishlistEntry:New(item, note)
    local o = {}
    setmetatable(o, self)

    o.item = item
    o.note = LimitNote(note)

    return o
end

local function NewWishlistEntry(item, note)
    return WishlistEntry:New(item, note)
end

function WishlistEntry:GetItemID()
    return self.item:GetItemID()
end

function WishlistEntry:GetItemLink()
    return self.item:GetItemLink()
end

function WishlistEntry:GetNote()
    return self.note
end

function WishlistEntry:SetNote(note)
    self.note = LimitNote(note)
end

function WishlistEntry:IsItemEmpty()
    return self.item:IsItemEmpty()
end

MWL.NewWishlistEntry = NewWishlistEntry

source/GUI.lua
local _,  MWL = ...

local AceGUI = LibStub("AceGUI-3.0")
local AceConfigRegistry = LibStub("AceConfigRegistry-3.0")
local AceConfigDialog = LibStub("AceConfigDialog-3.0")

local strsub20 = MWL.Utils.strsub20

local REGISTRY = "mywishlist_gui_registry"
local WISHLIST_REGISTRY = "mywishlist_gui_wishlist_registry"

local _, _, _, isElvUI = GetAddOnInfo("ElvUI")

local BASE_WIDTH_LOCKED    = 285 + (isElvUI and 10 or 0)
local BASE_WIDTH_UNLOCKED  = 535 + (isElvUI and 10 or 0)

-- local BASE_HEIGHT = 380

local GUI = {}

local options = {
    type = "group",
    args = {}
}
local wishlistOptions = {
    type = "group",
    args = {}
}

local function BuildWishlist_EntryIcon(order, icon, link)
    return {
        name = "",
        image = icon,
        type = "execute",
        func = (function() end),
        order = order,
        tooltipHyperlink = link,
        width = 0.25
    }, order + 1
end

local function BuildWishlist_EntryLink(order, link)
    return {
        name = link,
        type = "description",
        order = order,
        width = 1.25
    }, order + 1
end

local function GetEntryNote(entry)
    return entry:GetNote()
end

local function SetEntryNote(entry, note)
    if MWL.Manager:IsLocked() then return end
    entry:SetNote(note)
end

local function BuildWishlist_EntryNote(order, entry)
    return {
        name = "",
        type = "input",
        get = (function(i) return GetEntryNote(entry) end),
        set = (function(i, v) SetEntryNote(entry, v) end),
        order = order,
        width = 1
    }, order + 1
end

local function BuildWishlist_Up(order, slot, seqId, total)
    return {
        name = "",
        type = "execute",
        width = 0.15,
        image = "Interface\\Buttons\\UI-ScrollBar-ScrollUpButton-Up",
        hidden = (function() return (seqId > total) end),
        disabled = (function() return seqId == 1 end),
        func = function()
            MWL.Manager:MoveWishlistItemUp(slot, seqId)
            GUI:Refresh()
        end,
        order = order
    }, order + 1
end

local function BuildWishlist_Down(order, slot, seqId, total)
    return {
        name = "",
        type = "execute",
        hidden = (function() return (seqId > total) end),
        width = 0.15,
        image = "Interface\\Buttons\\UI-ScrollBar-ScrollDownButton-Up",
        disabled = (function() return seqId == total end),
        func = function()
            MWL.Manager:MoveWishlistItemDown(slot, seqId)
            GUI:Refresh()
        end,
        order = order
    }, order + 1
end

local function BuildWishlist_Delete(order, slot, seqId, total)
    return {
        name = "",
        type = "execute",
        hidden = (function() return (seqId > total) end),
        width = 0.15,
        image = "Interface\\Buttons\\UI-Panel-MinimizeButton-Up",
        func = function()
            MWL.Manager:RemoveWishlistItem(slot, seqId)
            GUI:Refresh()
        end,
        confirm = (function() return "Are you sure you want to remove item from wishlist?" end),
        order = order
    }, order + 1
end

local function BuildWishlist_EntryNoteReadOnly(order, entry)
    return {
        type = "description",
        name = (function(i) return MWL.Utils.ColorCodeText(GetEntryNote(entry),"44ee44") end),
        fontSize = "medium",
        order = order,
        width = 1.25
    }, order + 1
end

local function BuildWishlist_EntryPad(order)
    return {
        name = "",
        type = "description",
        width = 0.25,
        order = order
    }, order + 1
end


local lineNum = 0
local function BuildWishlistLine(order, entry, seqId, total)
    local itemId, _, _, _, icon, _, _ = GetItemInfoInstant(entry:GetItemID())
    if itemId then
        local link = entry:GetItemLink() or ("item:" .. tostring(itemId))
        local rowPrefix = "wlit" .. tostring(itemId) .. "ln" .. tostring(lineNum)

        wishlistOptions.args[rowPrefix .. "icon"  ], order = BuildWishlist_EntryIcon(order, icon, link)
        wishlistOptions.args[rowPrefix .. "link"  ], order = BuildWishlist_EntryLink(order, link)
        if MWL.Manager:IsLocked() then
            wishlistOptions.args[rowPrefix .. "pad"   ], order = BuildWishlist_EntryPad(order)
            wishlistOptions.args[rowPrefix .. "note"  ], order = BuildWishlist_EntryNoteReadOnly(order, entry)
        else
            wishlistOptions.args[rowPrefix .. "note"  ], order = BuildWishlist_EntryNote(order, entry)
            wishlistOptions.args[rowPrefix .. "up"    ], order = BuildWishlist_Up(order, MWL.Core.db.profile.selectedGuiSlot, seqId, total)
            wishlistOptions.args[rowPrefix .. "down"  ], order = BuildWishlist_Down(order, MWL.Core.db.profile.selectedGuiSlot, seqId, total)
            wishlistOptions.args[rowPrefix .. "delete"], order = BuildWishlist_Delete(order, MWL.Core.db.profile.selectedGuiSlot, seqId, total)
        end
        lineNum = lineNum + 1
    end
    return order + 1
end

local function UpdateOptions(self)
    options.args = {}
    wishlistOptions.args = {}
    local itemId, _, _, _, itemIcon, _, _ = GetItemInfoInstant(self.itemToAdd or 0)
    local icon = itemIcon or "Interface\\Icons\\INV_Misc_QuestionMark"
    local link = itemId and ("item:"..tostring(itemId)) or "item:0"

    local padding_width = 1.85
    if MWL.Manager:IsLocked() then
        padding_width = 0.35
    end

    options.args = {
        wishlist_slot_select = {
            name = "Slot",
            desc = "Select wishlist slot",
            type = "select",
            values = MWL.InternalSlots,
            sorting = MWL.InternalSlotsOrdered,
            set = function(i, v) MWL.Core.db.profile.selectedGuiSlot = v; self:Refresh() end,
            get = function(i) return MWL.Core.db.profile.selectedGuiSlot end,
            width = 1,
            order = 1
        },
        wishlist_lock = {
            -- name = (function(i) return MWL.Manager:IsLocked() and "Unlock" or "Lock" end),
            name = "",
            image = (function(i)
                if MWL.Manager:IsLocked() then
                    return "interface/addons/mywishlist/media/locked_3"
                else
                    return "interface/addons/mywishlist/media/unlocked_3"
                end
            end),
            type = "execute",
            func = (function(i)
                MWL.Manager:ToggleLock()
                self:Refresh()
            end),
            width = 0.25,
            order = 3
        },
        wishlist_padding_1 = {
            name = "",
            desc = "",
            type = "description",
            -- width = 1.75,
            width = padding_width,
            order = 2
        },
        add_header = {
            type = "header",
            name = "Add Item",
            width = "full",
            hidden = (function(i) return MWL.Manager:IsLocked() end),
            order = 4
        },
        wishlist_input_item_icon = {
            name = "",
            image = icon,
            type = "execute",
            func = (function() end),
            hidden = (function(i) return MWL.Manager:IsLocked() end),
            order = 5,
            tooltipHyperlink = link,
            width = 0.25
        },
        wishlist_input_item = {
            name = "",
            desc = "Input item link or item id to add to the wishlist. Item will be added to appropriate slot wishlist.",
            type = "input",
            set = (function(i, v)
                if tonumber(v) then
                    v = "item:" .. tostring(v)
                end
                local _itemId = MWL.Utils.GetItemIdFromLink(v)
                if _itemId and GetItemInfoInstant(_itemId) then
                    local item = Item:CreateFromItemID(_itemId)
                    if item:IsItemDataCached() then
                        self.itemToAdd = item:GetItemLink()
                        self:Refresh()
                    else
                        item:ContinueOnItemLoad(function()
                            self.itemToAdd = item:GetItemLink()
                            self:Refresh()
                        end)
                        -- self:Refresh()
                        self.itemToAdd = "item:" .. tostring(_itemId)
                    end
                end
            end),
            get = (function(i) return self.itemToAdd end),
            hidden = (function(i) return MWL.Manager:IsLocked() end),
            width = 1.25,
            order = 6
        },
        wishlist_input_note = {
            name = "",
            type = "input",
            get = (function(i) return self.note or "" end),
            set = (function(i, v) self.note = strsub20(tostring(v) or "") end),
            hidden = (function(i) return MWL.Manager:IsLocked() end),
            order = 7,
            width = 1
        },
        wishlist_add_item = {
            name = "Add",
            type = "execute",
            func = (function(i, v)
                MWL.Manager:AddItemByLink(self.itemToAdd, self.note)
                self.itemToAdd = ""
                self.note = ""
                self:Refresh()
            end),
            hidden = (function(i) return MWL.Manager:IsLocked() end),
            width = 0.6,
            order = 8
        },
        wishlist_header = {
            type = "header",
            name = "Wishlist",
            width = "full",
            order = 9
        },
    }
    local order = 50
    local wishlist = MWL.Manager:GetWishlist(MWL.Core.db.profile.selectedGuiSlot)
    for seqId, entry in ipairs(wishlist) do
        order = BuildWishlistLine(order, entry, seqId, #wishlist)
    end
end

local function CreateOptions(self)
    local OptionsGroup = AceGUI:Create("SimpleGroup")
    OptionsGroup:SetLayout("Flow")

    local WishlistOptionsScrollGroup = AceGUI:Create("ScrollFrame")

    local WishlistOptionsGroup = AceGUI:Create("SimpleGroup")
    WishlistOptionsGroup:SetLayout("Flow")

    WishlistOptionsScrollGroup:AddChild(WishlistOptionsGroup)

    UpdateOptions(self)
    AceConfigRegistry:RegisterOptionsTable(REGISTRY, options)
    AceConfigDialog:Open(REGISTRY, OptionsGroup)

    AceConfigRegistry:RegisterOptionsTable(WISHLIST_REGISTRY, wishlistOptions)
    AceConfigDialog:Open(WISHLIST_REGISTRY, WishlistOptionsGroup)

    self.OptionsGroup = OptionsGroup
    self.WishlistOptionsGroup = WishlistOptionsGroup
    self.WishlistOptionsScrollGroup = WishlistOptionsScrollGroup

    return OptionsGroup, WishlistOptionsScrollGroup
end

local function UpdateSize(self)
    if MWL.Manager:IsLocked() then
        self.window:SetWidth(BASE_WIDTH_LOCKED + 17)
        self.OptionsGroup:SetWidth(BASE_WIDTH_LOCKED + 17)
        self.WishlistOptionsGroup:SetWidth(BASE_WIDTH_LOCKED)
        self.WishlistOptionsScrollGroup:SetWidth(BASE_WIDTH_LOCKED + 17)

        self.WishlistOptionsScrollGroup:SetHeight(MWL.Core.db.profile.gui.height - 125 - 15)
    else
        self.window:SetWidth(BASE_WIDTH_UNLOCKED + 17)
        self.OptionsGroup:SetWidth(BASE_WIDTH_UNLOCKED + 17)
        self.WishlistOptionsGroup:SetWidth(BASE_WIDTH_UNLOCKED)
        self.WishlistOptionsScrollGroup:SetWidth(BASE_WIDTH_UNLOCKED + 17)

        self.WishlistOptionsScrollGroup:SetHeight(MWL.Core.db.profile.gui.height - 190 - 15)
    end
    self.OptionsGroup:DoLayout()
    self.WishlistOptionsGroup:DoLayout()
    self.WishlistOptionsScrollGroup:DoLayout()
end

local function CreateWindow(self)
    local f = AceGUI:Create("Window")

    f:SetTitle("My Wishlist")
    f:SetStatusText("")
    f:SetLayout("flow")
    f:SetHeight(MWL.Core.db.profile.gui.height)
    -- f:EnableResize(false)
    local originalOnHeightSet = f.OnHeightSet
    f.OnHeightSet = function(this, height)
        originalOnHeightSet(this, height)
        MWL.Core.db.profile.gui.height = height
        UpdateSize(self)
    end
    local originalOnMouseUp = f.title:GetScript("OnMouseUp")
    f.title:SetScript("OnMouseUp", function(...)
        originalOnMouseUp(...)
        MWL.Core.db.profile.gui.location = { f:GetPoint() }
    end)

    if MWL.Core.db.profile.gui.location then
        f:ClearAllPoints()
        f:SetPoint(MWL.Core.db.profile.gui.location[3], MWL.Core.db.profile.gui.location[4], MWL.Core.db.profile.gui.location[5])
    end

    local op, sc = CreateOptions(self)
    f:AddChild(op)
    f:AddChild(sc)
    self.window = f

    self:Hide()
end

function GUI:Initialize()
    CreateWindow(self)
    UpdateSize(self)
end

function GUI:Refresh()
    UpdateOptions(self)
    AceConfigRegistry:RegisterOptionsTable(REGISTRY, options)
    AceConfigDialog:Open(REGISTRY, self.OptionsGroup)
    AceConfigRegistry:RegisterOptionsTable(WISHLIST_REGISTRY, wishlistOptions)
    AceConfigDialog:Open(WISHLIST_REGISTRY, self.WishlistOptionsGroup)
    UpdateSize(self)
end

function GUI:Show()
    self:Refresh()
    self.window:Show()
end

function GUI:Hide()
    self.window:Hide()
end

function GUI:Toggle()
    if self.window:IsVisible() then
        self.window:Hide()
    else
        self.window:Show()
    end
end


MWL.GUI = GUI